lib/factory_bot/strategy_syntax_method_registrar.rb

```diff
@@ -33,6 +33,21 @@ def define_list_strategy_method
       end
     end
 
+    def define_list_with_index_strategy_method
```

Comment:
I think your tests are failing because you aren't calling this anywhere. You would need to call it inside of `define_strategy_methods`.

That said, I am leaning toward extending the existing `*_list` method rather than introducing a new `*_list_with_index` method. We would extend the existing method by doing an arity check to decide whether to pass along the index. This is more in line with what we have done in other place in factory bot (e.g. [passing the evaluator to dynamic attributes](https://github.com/thoughtbot/factory_bot/blob/6d8bea1670ef10daf0f15680288334916e031430/lib/factory_bot/attribute/dynamic.rb#L14-L17), [passing the instance to to_create](https://github.com/thoughtbot/factory_bot/blob/c22c9ab0524a00ede1f09a1c996ed77dfc1afa73/lib/factory_bot/evaluation.rb#L16-L19), or [passing both the instance and the evaluator to callbacks](https://github.com/thoughtbot/factory_bot/blob/f82e40c8c50c538c0fba0362c0f6b7f6f96f9a3c/lib/factory_bot/callback.rb#L12-L16))

I had played around with this in https://github.com/thoughtbot/factory_bot/commit/6996d8287510aa2f415c8ddf5745a1022fdb837d. I don't remember how far I got, but it should give you the general idea of what I was thinking about.

Reply:
Thanks for the feedback! Extending the method sounds great, I gave it a try in c4443cb1d905552a39475e3acf2e26040c58f9c6, it looks like it's working well.